Add test for winstore loading

nhorman opened the pull request:
Add a rudimentary test to ensure that we can load the windows store

##### Checklist
- [x] tests are added or updated

mattcaswell commented on test/sslapitest.c:
```diff
@@ -13093,6 +13093,24 @@ static int test_quic_tls_early_data(void)
+
+    if (!TEST_ptr(ctx))
+        goto err;
+    if (!TEST_int_eq(SSL_CTX_load_verify_store(ctx, "org.openssl.winstore:"), 1))
```
It seems strange to be testing this via SSL. Why not use the `OSSL_STORE_*` APIs directly? Obviously not in sslapitest.c though - perhaps ossl_store_test.c might be a more obvious place.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -13093,6 +13093,24 @@ static int test_quic_tls_early_data(void)
+    return ret;
+}
+
+#endif
```
I tried taking out the `ifdef OPENSSL_SYS_WINDOWS` and running this on Linux. The test passed. I don't think this does what you think it does. This doesn't actually load the winstore. It just stores away the URI in an internal stack ready for later use.

nhorman commented:
reworked the test and moved it to ossl_store_test.c.  

Waiving style as the test commands will either overflow the 100 column limit or use invalid spacing.  The latter is the style of the file it appears.

Sashan reviewed (approved):
looks good to me. but CI is relevant.

mattcaswell commented on test/ossl_store_test.c:
```diff
@@ -58,6 +58,28 @@ static int test_store_open(void)
+    UI_METHOD *ui_method = NULL;
+
+    ret = TEST_ptr(search = OSSL_STORE_SEARCH_by_alias("nothing"))
+          && TEST_ptr(ui_method= UI_create_method("DummyUI"))
```
Nit
```suggestion
          && TEST_ptr(ui_method = UI_create_method("DummyUI"))
```

mattcaswell reviewed (approved):
Approved assuming the nit is fixed. Have you tried this on Windows without your winstore fix and confirmed that the test fails?

t8m reviewed (dismissed):
Approved if CI passes. It should go also to 3.5 branch at least.

nhorman commented:
@Sashan @mattcaswell test updated to actually look for a real certificate we know is there (the MS root authority) so that we exercise the winstore search/load code properly.  Please review/reapprove

Sashan commented:
it's still not quite right. I have no clue how this PKI is supposed to work. My observation is based on my attempt to debug this stuff. The function we attempt to test (`setup_decoder()`) is not being called. This is what I'm seeing while trying to debug it.

The first screenshot shows we call to `winstore_open()` on behalf of `OSSL_STORE_open_ex()`. We call it on behalf of `OSSL_STORE_find()` we have in testcase.
![ScreenShot-01](https://github.com/user-attachments/assets/208bdc7d-bf92-42d0-be05-6fb9826bc480)


The function opens `ctx->win_store` successfully. The function also calls `wintore_win_reset()` which sets `ctx->state` to `STATE_IDLE`.

Later we call to `winstore_win_advance()`  (still on behalf of `OSSL_STORE_find()`) as seen on next screenshot:
![ScreenShot-02](https://github.com/user-attachments/assets/669ff939-8ef6-4739-9ee7-5e710678c82e)

The thing is the call to `CertFindCertificateInStore()` fails. It returns `NULL` which makes the function to set `ctx->state` to STATE_EOF.  I have no idea why it fails. I will discuss that later.

Note failing to find certificate does not trigger error in `OSSL_STORE_find()` path `winstore_set_ctx_params()` returns 1 (success indication) anyway.

The test now calls `OSSL_store_load()` to load desired certificate as seen on next screenshot here:
![ScreenShot-03](https://github.com/user-attachments/assets/481f58b8-bde3-4fcf-b17b-17e5182f76f4)

The function bails out at line 426 because the winstore's internal ctx state is at `STATE_EOF`.

The test attempts to load windows certificate with subject as follows:
```
CN: Microsoft Root Certificate Authority 2011
O: Microsoft Corporation
L: Redmond
ST: Washington
C: US
```

Such certificate seems to exist on my windows as seen on screen shot here:
![ScreenShot-04](https://github.com/user-attachments/assets/2a39906f-2387-495a-b41b-fa887f33e6fa)

The only difference is that `Subject` in certificate has item `S: Washington` while test case uses `ST: Washington`. This made me thinking to apply small tweak to the test:
```
diff --git a/test/ossl_store_test.c b/test/ossl_store_test.c
index 5da55b098a..5f5077101b 100644
--- a/test/ossl_store_test.c
+++ b/test/ossl_store_test.c
@@ -77,7 +77,7 @@ static int test_store_open_winstore(void)
         && TEST_true(X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, (unsigned char *)"Microsoft Root Certificate Authority 2011", -1, -1, 0))
         && TEST_true(X509_NAME_add_entry_by_txt(name, "O", MBSTRING_ASC, (unsigned char *)"Microsoft Corporation", -1, -1, 0))
         && TEST_true(X509_NAME_add_entry_by_txt(name, "L", MBSTRING_ASC, (unsigned char *)"Redmond", -1, -1, 0))
-        && TEST_true(X509_NAME_add_entry_by_txt(name, "ST", MBSTRING_ASC, (unsigned char *)"Washington", -1, -1, 0))
+        && TEST_true(X509_NAME_add_entry_by_txt(name, "S", MBSTRING_ASC, (unsigned char *)"Washington", -1, -1, 0))
         && TEST_true(X509_NAME_add_entry_by_txt(name, "C", MBSTRING_ASC, (unsigned char *)"US", -1, -1, 0))
         && TEST_ptr(search = OSSL_STORE_SEARCH_by_name(name))
         && TEST_ptr(ui_method = UI_create_method("DummyUI"))
```

However once tweak above is in place the test starts to fail with error as follows
```
# The results of this test will end up in test-runs\test_ossl_store
1..1
    # Subtest: ..\..\test\ossl_store_test.exe
    1..5
    ok 1 - test_store_open
    # ERROR: (bool) 'X509_NAME_add_entry_by_txt(name, "S", MBSTRING_ASC, (unsigned char *)"Washington", -1, -1, 0) == true' failed @ test\ossl_store_test.c:80
    # false
    # 20140000:error:04000067:object identifier routines:OBJ_txt2obj:unknown object name:crypto\objects\obj_dat.c:438:
    # 20140000:error:05800077:x509 certificate routines:X509_NAME_ENTRY_create_by_txt:invalid field name:crypto\x509\x509name.c:252:name=S
    # OPENSSL_TEST_RAND_SEED=1745140272
    not ok 2 - test_store_open_winstore

        ok 3 - test_store_search_by_key_fingerprint_fail
```

I'm not sure what needs to be done to get the test working. Perhaps we also need to fix X509_NAME_*()` machinery to allow for 'S' instead of 'ST'. But I'm not qualified enough here to judge as I know nothing about PKI/certs.

Sashan commented:
I've also tried to search for different certificate.
```
diff --git a/test/ossl_store_test.c b/test/ossl_store_test.c
index 5da55b098a..a5a533db70 100644
--- a/test/ossl_store_test.c
+++ b/test/ossl_store_test.c
@@ -74,11 +74,9 @@ static int test_store_open_winstore(void)
      * subject name, as winstore only allows searches by that method
      */
     ret = TEST_ptr(name)
-        && TEST_true(X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, (unsigned char *)"Microsoft Root Certificate Authority 2011", -1, -1, 0))
-        && TEST_true(X509_NAME_add_entry_by_txt(name, "O", MBSTRING_ASC, (unsigned char *)"Microsoft Corporation", -1, -1, 0))
-        && TEST_true(X509_NAME_add_entry_by_txt(name, "L", MBSTRING_ASC, (unsigned char *)"Redmond", -1, -1, 0))
-        && TEST_true(X509_NAME_add_entry_by_txt(name, "ST", MBSTRING_ASC, (unsigned char *)"Washington", -1, -1, 0))
-        && TEST_true(X509_NAME_add_entry_by_txt(name, "C", MBSTRING_ASC, (unsigned char *)"US", -1, -1, 0))
+        && TEST_true(X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, (unsigned char *)"Microsoft Root Certificate Authority", -1, -1, 0))
+        && TEST_true(X509_NAME_add_entry_by_txt(name, "DC", MBSTRING_ASC, (unsigned char *)"microsoft", -1, -1, 0))
+        && TEST_true(X509_NAME_add_entry_by_txt(name, "DC", MBSTRING_ASC, (unsigned char *)"com", -1, -1, 0))
         && TEST_ptr(search = OSSL_STORE_SEARCH_by_name(name))
         && TEST_ptr(ui_method = UI_create_method("DummyUI"))
         && TEST_ptr(sctx = OSSL_STORE_open_ex("org.openssl.winstore:", NULL,
```
The tweak above is accepted by `X509_add_*()` and should match certificate shown on screenshot here:
![ScreenShot-05](https://github.com/user-attachments/assets/f7e23cbd-a879-4816-a337-4c3d9013dbdf)

With diff above I'm seeing the same behavior as with the change presented in PR. things fail in `CertFindCertificateInStore()` which returns `NULL`.  I'm out of ideas how to debug it further at the moment.

nhorman commented:
I noted the S/ST discrepancy as well, looking at the Known root certificate in CMM.  The thing I noted was that CMM displays the S attribute as the same NID as ST in openssl - i.e. SN_stateOrProvinceName as defined in openssl as "ST" and maps to the definition of NID_stateOrProvinceName (16), which is as far as I can see the same NID in windows, using the short name "S".  So I think thats ok.

As for why you're not finding the certificate my local system successfully finds, I'm not sure.  My first though is that you may be using an older version of my branch in which I had forgotten to call OSSL_STORE_find prior to calling OSSL_STORE_load, but given that you're getting into winstore_win_advance, that doesn't seem likely.

Perhaps the best thing to do at this point is figure out getting shared access to your vm on proxmox, so I can compare certificates on your system to the store on my local vm.  Perhaps the certificate we are trying to find isn't as widely available as I first assumed.

Sashan commented:
I will try to figure out RDP. meanwhile I've crafted a draft [draft](https://github.com/openssl/openssl/pull/27453) see how it will work on CI. draft PR expects your test to fail.

Sashan commented:
> I will try to figure out RDP. meanwhile I've crafted a draft [draft](https://github.com/openssl/openssl/pull/27453) see how it will work on CI. draft PR expects your test to fail.

The test [fails](https://github.com/openssl/openssl/actions/runs/14570130025/job/40865761911?pr=27453#step:10:297), but it fails because I've forgot to revert my own test changes. It needs this one-liner tweak to turn it back to the code I got from Neil:
```
diff --git a/test/ossl_store_test.c b/test/ossl_store_test.c
index 5f5077101b..5da55b098a 100644
--- a/test/ossl_store_test.c
+++ b/test/ossl_store_test.c
@@ -77,7 +77,7 @@ static int test_store_open_winstore(void)
         && TEST_true(X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, (unsigned char *)"Microsoft Root Certificate Authority 2011", -1, -1, 0))
         && TEST_true(X509_NAME_add_entry_by_txt(name, "O", MBSTRING_ASC, (unsigned char *)"Microsoft Corporation", -1, -1, 0))
         && TEST_true(X509_NAME_add_entry_by_txt(name, "L", MBSTRING_ASC, (unsigned char *)"Redmond", -1, -1, 0))
-        && TEST_true(X509_NAME_add_entry_by_txt(name, "S", MBSTRING_ASC, (unsigned char *)"Washington", -1, -1, 0))
+        && TEST_true(X509_NAME_add_entry_by_txt(name, "ST", MBSTRING_ASC, (unsigned char *)"Washington", -1, -1, 0))
         && TEST_true(X509_NAME_add_entry_by_txt(name, "C", MBSTRING_ASC, (unsigned char *)"US", -1, -1, 0))
         && TEST_ptr(search = OSSL_STORE_SEARCH_by_name(name))
         && TEST_ptr(ui_method = UI_create_method("DummyUI"))
```

so time for one more try.

Sashan commented:
with fix-up in the windows-2022-plain tests succeed while they should fail in fact (https://github.com/openssl/openssl/actions/runs/14571709592/job/40870069066?pr=27453).

nhorman commented:
@Sashan Why should it fail?  The certificate is in the store, so we find it, no?

Sashan commented:
> @Sashan Why should it fail? The certificate is in the store, so we find it, no?

I expect test to fail, because the PR (https://github.com/openssl/openssl/pull/27453) also reverts the [fix](https://github.com/openssl/openssl/pull/27453/commits/27ae55c797831d4808b400c1ea9533b2d1faf8e9) for crash

nhorman commented:
@Sashan I think I understand why the test isn't failing when you back out the previous fix.

The key is roughly in @vdukhovni comment [here](https://github.com/openssl/openssl/pull/27384#issuecomment-2802282184)

The initial problem wasn't that we had a NULL pointer that was erroneously unset (i.e. we just didn't set something up correctly).  It was that, prior to his previous changes, the OSSL_DECODER_INSTANCE returned from ossl_decoder_instance_new was equivalent between all providers.  With his changes, we need to have the OSSL_DECODER_INSTANCE decoderctx member set with the value returned from decoder->newctx.  Prior to the fixing change we just passed ctx->provctx, which is the wrong type.  As a result we were dereferencing memory that held  provctx type, but the code that uses is was expected to hold an opaque type defined by the decoder.  Because they're all void pointers, there was/is no typechecking, and so the winstore provider was assuming it was working with a winstore_ctx_st struct, when it was really a PROV_CTX struct.

Thats a mess, but it is what it is.  We should look at improving that  at some point. 

Point being that, because we weren't matching types properly, I'm not sure that we will always get a crash.  We might just read garbage memory, but if the pointer is valid, its valid.

Sashan commented:
The main issue which needs to get solved here is to make sure the function `setup_decoder()` (the fixed function) gets called by test. And this is not the case. The `setup_decoder()` is never reached by this test case. The `OSSL_STORE_load()` which should call to `setup_decoder()` bails out as soon as it is entered because `OSSL_STORE_eof()` is true.

nhorman commented:
@Sashan ok, thank you I understand now.  I see now that I made two mistakes:
1) I'm testing in the wrong sense on the info pointer in the ossl_store_test.  I should be doing TEST_ptr rather than !TEST_ptr.  Not sure how I let that in there
2) The OSSL_STORE_find call is returning 0 certificates for the subject we're looking for.  Unsure why thats happening, as I see the certificate there, and I could have sworn I had that working previously. 

I'll look into this more.

nhorman commented:
@Sashan @mattcaswell could you take another look at this please.  Based on @Sashan's findings, I realized this test was wrong, and fixing it caused a lookup failure.

For some reason, generating an X509_NAME with the same subject as a known certificate in the store isn't causing a successful lookup.  After about a week I've been unable to determine why, but If I export the cert I'm looking for to a hex array in the test, convert it, via d2i, and extract the subject name from it, the lookup works.  We probably need to look into that further, but I think this solution gives us a sufficient first test to confirm winstore works.  And it fails If I revert the prior winstore fix.

vdukhovni commented on test/ossl_store_test.c:
```diff
@@ -58,6 +58,165 @@ static int test_store_open(void)
+  0x0b, 0x05, 0x00, 0x03, 0x82, 0x02, 0x01, 0x00, 0x7f, 0x72, 0xcf, 0x0f, 0xb7,
+  0xc5, 0x15, 0xdb, 0x9b, 0xc0, 0x49, 0xca, 0x26, 0x5b, 0xfe, 0x9e, 0x13, 0xe6,
+  0xd3, 0xf0, 0xd2, 0xdb, 0x97, 0x5f, 0xf2, 0x4b, 0x3f, 0x4d, 0xb3, 0xae, 0x19,
+  0xae, 0xed, 0xd7, 0x97, 0xa0, 0xac, 0xef, 0xa9, 0x3a, 0xa3, 0xc2, 0x41, 0xb0,
+  0xe5, 0xb8, 0x91, 0x9e, 0x13, 0x81, 0x24, 0x03, 0xe6, 0x09, 0xfd, 0x3f, 0x57,
+  0x40, 0x39, 0x21, 0x24, 0x56, 0xd1, 0x10, 0x2f, 0x4b, 0x40, 0xa9, 0x36, 0x86,
+  0x4b, 0xb4, 0x53, 0x57, 0x9a, 0xfb, 0xf1, 0x7e, 0x89, 0x8f, 0x11, 0xfe, 0x18,
+  0x6c, 0x51, 0xaa, 0xe8, 0xed, 0x09, 0x95, 0xb5, 0xe5, 0x71, 0xc9, 0xa1, 0xe9,
+  0x87, 0x75, 0xa6, 0x15, 0x7f, 0xc9, 0x7e, 0x37, 0x54, 0x5e, 0x74, 0x93, 0xc5,
+  0xc3, 0x67, 0xcc, 0x0d, 0x4f, 0x6b, 0xa8, 0x17, 0x0c, 0x6d, 0x08, 0x92, 0x7e,
+  0x8b, 0xdd, 0x81, 0xaa, 0x2d, 0x70, 0x21, 0xc3, 0x3d, 0x06, 0x14, 0xbb, 0xbf,
+  0x24, 0x5e, 0xa7, 0x84, 0xd7, 0x3f, 0x0f, 0x21, 0x22, 0xbd, 0x4b, 0x00, 0x06,
+  0xdb, 0x97, 0x1c, 0xd8, 0x5e, 0xd4, 0xc5, 0x0b, 0x5c, 0x87, 0x6e, 0x50, 0xa4,
+  0xe8, 0xc3, 0x38, 0xa4, 0xfb, 0xcb, 0x2c, 0xc5, 0x92, 0x66, 0x9b, 0x85, 0x5e,
+  0xcb, 0x7a, 0x6c, 0x93, 0x7c, 0x80, 0x29, 0x58, 0x5b, 0x57, 0xb5, 0x40, 0x69,
+  0xba, 0x08, 0x79, 0xa6, 0x64, 0x62, 0x15, 0x9d, 0x87, 0x96, 0x45, 0xb5, 0x66,
+  0x23, 0x20, 0x03, 0x8b, 0x1c, 0x73, 0xa0, 0xd3, 0xa2, 0x79, 0x33, 0xe0, 0x50,
+  0x59, 0x86, 0xdb, 0x2f, 0xe5, 0x02, 0x25, 0xea, 0x73, 0x2a, 0x9f, 0x00, 0x14,
+  0xc8, 0x36, 0xc7, 0x92, 0x3b, 0xe9, 0x4e, 0x00, 0xec, 0xd8, 0x56, 0x09, 0xb9,
+  0x33, 0x49, 0x12, 0xd2, 0x54, 0x0b, 0x01, 0xab, 0xac, 0x47, 0xb6, 0x91, 0x29,
+  0x7d, 0x4c, 0xb4, 0x75, 0x80, 0x52, 0x01, 0xe8, 0xca, 0x82, 0xf6, 0x9f, 0xcc,
+  0xac, 0x9c, 0x8f, 0x17, 0xea, 0x2f, 0x26, 0xb0, 0xab, 0x72, 0xac, 0x0b, 0xfe,
+  0x9e, 0x51, 0x1e, 0xc7, 0x43, 0x55, 0x67, 0x4f, 0x51, 0xb3, 0x57, 0xd6, 0xb6,
+  0xec, 0xee, 0x52, 0xb7, 0x3a, 0xe9, 0x4e, 0xe1, 0xd7, 0x81, 0x88, 0xbc, 0x4f,
+  0x8e, 0x75, 0xbb, 0x4b, 0xa8, 0xf0, 0x35, 0xaa, 0x26, 0xd4, 0x67, 0x67, 0x49,
+  0xb2, 0x70, 0x4c, 0x3b, 0x93, 0xdc, 0x1d, 0xdf, 0x78, 0x90, 0x86, 0x72, 0xb2,
+  0x38, 0xa4, 0xd1, 0xdc, 0x92, 0x4d, 0xc9, 0x58, 0xeb, 0x2b, 0x12, 0x5c, 0xd4,
+  0x3b, 0xae, 0x8c, 0x6b, 0xb0, 0x83, 0xe5, 0x01, 0x3f, 0xf8, 0x09, 0x32, 0xf6,
+  0x93, 0x35, 0x34, 0x22, 0xaf, 0xdd, 0x37, 0x0d, 0x77, 0x09, 0x80, 0x2b, 0xcd,
+  0x48, 0x00, 0xf1, 0x8c, 0x99, 0x19, 0x47, 0x05, 0x01, 0xe9, 0xd1, 0xbf, 0xd1,
+  0x4e, 0xd0, 0xe6, 0x28, 0x43, 0x37, 0x99, 0xa4, 0x0a, 0x4a, 0x08, 0xd9, 0x9a,
+  0x71, 0x73, 0xd2, 0xaa, 0xcd, 0x31, 0x13, 0x63, 0x76, 0xa1, 0x37, 0x6f, 0x92,
+  0x38, 0x1e, 0x7d, 0x12, 0x3c, 0x66, 0x32, 0xe7, 0xcb, 0x6d, 0xe1, 0xfc, 0x52,
+  0x89, 0xdd, 0xca, 0xd6, 0x66, 0x05, 0x9a, 0x96, 0x61, 0xbe, 0xa2, 0x28, 0xc7,
+  0x1c, 0xa3, 0xa7, 0x36, 0x50, 0x3c, 0x3a, 0xa4, 0xdf, 0x4a, 0x6e, 0xe6, 0x87,
+  0x3b, 0xce, 0xeb, 0xf0, 0xe0, 0x81, 0x37, 0x9d, 0x13, 0x3c, 0x52, 0x8e, 0xbd,
+  0xb9, 0x1d, 0x34, 0xc6, 0x1d, 0xd5, 0x0a, 0x6a, 0x3d, 0x98, 0x29, 0x70, 0x8c,
+  0x89, 0x2a, 0xd1, 0xab, 0x82, 0x10, 0x48, 0x1f, 0xdc, 0xf4, 0xef, 0xa5, 0xc5,
+  0xbb, 0x55, 0x1a, 0x38, 0x63, 0x84, 0x4e, 0xb7, 0x6c, 0xad, 0x95, 0x54, 0xec,
+  0x65, 0x22, 0x10, 0x49, 0x17, 0xb8, 0xc0, 0x1e, 0xc7, 0x0f, 0xac, 0x54, 0x47
```
Since all we need is a certificate with a particular name, this is much shorter, with the desired name as issuer.
```suggestion
  0x30, 0x82, 0x01, 0x3b, 0x30, 0x81, 0xee, 0xa0, 0x03, 0x02, 0x01, 0x02,
  0x02, 0x01, 0x00, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x30, 0x81,
  0x88, 0x31, 0x0b, 0x30, 0x09, 0x06, 0x03, 0x55, 0x04, 0x06, 0x13, 0x02,
  0x55, 0x53, 0x31, 0x13, 0x30, 0x11, 0x06, 0x03, 0x55, 0x04, 0x08, 0x13,
  0x0a, 0x57, 0x61, 0x73, 0x68, 0x69, 0x6e, 0x67, 0x74, 0x6f, 0x6e, 0x31,
  0x10, 0x30, 0x0e, 0x06, 0x03, 0x55, 0x04, 0x07, 0x13, 0x07, 0x52, 0x65,
  0x64, 0x6d, 0x6f, 0x6e, 0x64, 0x31, 0x1e, 0x30, 0x1c, 0x06, 0x03, 0x55,
  0x04, 0x0a, 0x13, 0x15, 0x4d, 0x69, 0x63, 0x72, 0x6f, 0x73, 0x6f, 0x66,
  0x74, 0x20, 0x43, 0x6f, 0x72, 0x70, 0x6f, 0x72, 0x61, 0x74, 0x69, 0x6f,
  0x6e, 0x31, 0x32, 0x30, 0x30, 0x06, 0x03, 0x55, 0x04, 0x03, 0x13, 0x29,
  0x4d, 0x69, 0x63, 0x72, 0x6f, 0x73, 0x6f, 0x66, 0x74, 0x20, 0x52, 0x6f,
  0x6f, 0x74, 0x20, 0x43, 0x65, 0x72, 0x74, 0x69, 0x66, 0x69, 0x63, 0x61,
  0x74, 0x65, 0x20, 0x41, 0x75, 0x74, 0x68, 0x6f, 0x72, 0x69, 0x74, 0x79,
  0x20, 0x32, 0x30, 0x31, 0x31, 0x30, 0x20, 0x17, 0x0d, 0x32, 0x35, 0x30,
  0x34, 0x32, 0x39, 0x30, 0x35, 0x34, 0x33, 0x33, 0x30, 0x5a, 0x18, 0x0f,
  0x32, 0x31, 0x32, 0x35, 0x30, 0x34, 0x32, 0x39, 0x30, 0x35, 0x34, 0x33,
  0x33, 0x30, 0x5a, 0x30, 0x00, 0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b,
  0x65, 0x70, 0x03, 0x21, 0x00, 0x21, 0x86, 0x38, 0x28, 0x7d, 0xad, 0x35,
  0xad, 0xc3, 0x91, 0x07, 0x41, 0x65, 0x02, 0xe1, 0x72, 0x79, 0xf4, 0x0d,
  0xd5, 0xe2, 0xbb, 0xba, 0xd8, 0x81, 0x06, 0xad, 0xa0, 0x7b, 0x2a, 0x41,
  0x09, 0xa3, 0x02, 0x30, 0x00, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70,
  0x03, 0x41, 0x00, 0x2e, 0xbb, 0xb9, 0x55, 0x59, 0x24, 0xb0, 0xdb, 0x18,
  0xd0, 0xb4, 0xa6, 0x8b, 0xa8, 0xbf, 0x4c, 0xbd, 0x83, 0x27, 0xe7, 0xc2,
  0xc8, 0x13, 0xc2, 0x81, 0xbe, 0xdf, 0x04, 0x7b, 0x09, 0x87, 0x08, 0x53,
  0x97, 0xd6, 0x7c, 0xb7, 0x6c, 0xa6, 0x78, 0x63, 0x3c, 0x27, 0x29, 0x81,
  0x85, 0xd0, 0x73, 0x92, 0x2d, 0x5c, 0x33, 0x8e, 0x88, 0x18, 0x9b, 0x1c,
  0xcf, 0xc3, 0x4a, 0xf0, 0x82, 0x05, 0x05
```

nhorman replied:
have you tested this?

vdukhovni commented on test/ossl_store_test.c:
```diff
@@ -58,6 +58,165 @@ static int test_store_open(void)
+     * subject name, as winstore only allows searches by that method
+     */
+    ret = TEST_ptr(testcert = d2i_X509(NULL, &certptr, TEST_CERT_LEN))
+        && TEST_ptr(search = OSSL_STORE_SEARCH_by_name(X509_get_subject_name(testcert)))
```
```suggestion
        && TEST_ptr(search = OSSL_STORE_SEARCH_by_name(X509_get_issuer_name(testcert)))
```

vdukhovni reviewed (approved):
Optional, use a much more compact certificate to embed the name as issuer.

vdukhovni reviewed (approved):
With a suggestion to update the comment.

Sashan reviewed (approved):
Looks good to me. I did test it. The test crashes when fix to winstore is not present. The terminal output reads as follows:
```
C:\Strawberry\perl\bin\perl.exe" ".\test\run_tests.pl" test_ossl_store"
00-prep_fipsmodule_cnf.t .. skipped: FIPS module config file only supported in a fips build
Files=1, Tests=0,  1 wallclock secs ( 0.00 usr +  0.01 sys =  0.01 CPU)
Result: NOTESTS
66-test_ossl_store.t ..
Can't spawn "C:\Windows\system32\cmd.exe": No such file or directory at ..\..\util\wrap.pl line 64.
C:\Strawberry\perl\bin\perl.exe ..\..\util\wrap.pl ..\..\test\ossl_store_test.exe -dir ..\..\test -in testrsa.pem -sm2 certs\sm2-root.crt -data ..\..\test\recipes\66-test_ossl_store_data => 255
not ok 1
66-test_ossl_store.t .. 1/? ----------------------------------------------------
#   Failed test at test\recipes\66-test_ossl_store.t line 20.
66-test_ossl_store.t .. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
 
Test Summary Report
-------------------
66-test_ossl_store.t (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=1, Tests=1, 62 wallclock secs ( 0.02 usr +  0.00 sys =  0.02 CPU)
Result: FAIL
NMAKE : fatal error U1077: 'cmd' : return code '0x1'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
 
C:\Users\OpenSSL\work.openssl\openssl.master>nmake test TESTS=test_ossl_store
```
It feels like a NULL pointer dereference (see image attached)
![Crash-01](https://github.com/user-attachments/assets/b146a7e1-bb8d-4d88-ad8d-57f255a1ea0f)

Test passes on openssl where fix is present:
```
C:\Strawberry\perl\bin\perl.exe" ".\test\run_tests.pl" test_ossl_store"
00-prep_fipsmodule_cnf.t .. skipped: FIPS module config file only supported in a fips build
Files=1, Tests=0,  0 wallclock secs ( 0.00 usr +  0.03 sys =  0.03 CPU)
Result: NOTESTS
66-test_ossl_store.t .. ok
All tests successful.
Files=1, Tests=1,  1 wallclock secs ( 0.00 usr +  0.00 sys =  0.00 CPU)
Result: PASS
```
 Well done. Thanks.

openssl-machine commented:
This pull request is ready to merge
